gems/aws-sdk-core/lib/aws-sdk-core/plugins/regional_endpoint.rb

```diff
@@ -93,6 +96,66 @@ def after_initialize(client)
         client.config.sigv4a_signing_region_set = region_set
       end
 
+      private
+
+      def initialize_default_endpoint(client)
+        client_module = Object.const_get(client.class.name.rpartition('::').first)
+        if client.config.respond_to?(:endpoint_provider) &&
+          client_module.const_defined?(:EndpointParameters) &&
+          (param_class = client_module.const_get(:EndpointParameters)) &&
+          (endpoint_provider = client.config.endpoint_provider)
+
+          params = build_parameters(param_class.new, client.config)
+          begin
+            endpoint = endpoint_provider.resolve_endpoint(params)
+            client.config.endpoint = endpoint.url
+          rescue ArgumentError
+            # fallback to legacy
+            client.config.endpoint = resolve_legacy_endpoint(client.config)
+          end
+        else
+          # fallback to legacy
+          client.config.endpoint = resolve_legacy_endpoint(client.config)
+        end
+      end
+
+      def build_parameters(params, config)
+        # generic parameters set for most (but not all) services
+        params.region = config.region if params.respond_to?(:region=)
+        params.use_dual_stack = config.use_dualstack_endpoint if params.respond_to?(:use_dual_stack=)
+        params.use_fips = config.use_fips_endpoint if params.respond_to?(:use_fips=)
+
+        # service specific
+        if config.respond_to?(:sts_regional_endpoints) && params.respond_to?(:use_global_endpoint=)
+          params.use_global_endpoint = (config.sts_regional_endpoints == 'legacy')
```

Comment:
Do these not have defaults in the params class? Maybe the params class can take config to construct defaults.

Reply:
These do have defaults. But this is required to get the "correct" values based on customers config.

Reply:
I'll look into taking config when constructing params. That would be a codegen change that would regenerate every service and I'm not sure want to couple params to config since we have a separate way to construct it (per operation)